qsc/grad_B_tensor.py

```diff
@@ -1177,3 +1190,233 @@ def calculate_grad_grad_B_tensor(self, two_ways=False):
+    b = self.binormal_cylindrical.transpose()
+    Bbar = self.Bbar
+    B0 = self.B0
+    sG = self.sG
+    G0 = self.G0
+    X1c = self.X1c
+    X1s = self.X1s
+    Y1c = self.Y1c
+    Y1s = self.Y1s
+    d_l_d_varphi = self.d_l_d_varphi
+    curvature = self.curvature
+    torsion = self.torsion
+    iotaN = self.iotaN
+    d_X1c_d_varphi = self.d_X1c_d_varphi
+    d_X1s_d_varphi = self.d_X1s_d_varphi
+    d_Y1s_d_varphi = self.d_Y1s_d_varphi
+    d_Y1c_d_varphi = self.d_Y1c_d_varphi
+
+    factor = B0*B0/Bbar/d_l_d_varphi
+    grad_B_vector_nn = factor * (d_X1c_d_varphi * Y1s - d_X1s_d_varphi * Y1c + iotaN * (X1s * Y1s + X1c * Y1c))
+    grad_B_vector_bn = factor * (X1c * d_X1s_d_varphi - X1s * d_X1c_d_varphi - sG * Bbar * d_l_d_varphi * torsion / B0 - iotaN * (X1s * X1s + X1c * X1c))
+    grad_B_vector_nb = factor * (d_Y1c_d_varphi * Y1s - d_Y1s_d_varphi * Y1c + sG * Bbar * d_l_d_varphi * torsion / B0 + iotaN * (Y1s * Y1s + Y1c * Y1c))
+    grad_B_vector_bb = factor * (X1c * d_Y1s_d_varphi - X1s * d_Y1c_d_varphi - iotaN * (X1s * Y1s + X1c * Y1c))
+    grad_B_vector_tn = sG * B0 * curvature
+    grad_B_vector_nt = sG * B0 * curvature
+    grad_B_vector_tt = 0#sG * np.matmul(self.d_d_varphi,B0)/d_l_d_varphi
+
+    grad_B_vector_cylindrical = np.array([[
+                                  grad_B_vector_nn * n[i] * n[j] \
+                                + grad_B_vector_bn * b[i] * n[j] + grad_B_vector_nb * n[i] * b[j] \
+                                + grad_B_vector_bb * b[i] * b[j] \
+                                + grad_B_vector_tn * t[i] * n[j] + grad_B_vector_nt * n[i] * t[j] \
+                                + grad_B_vector_tt * t[i] * t[j]
+                                for i in range(3)] for j in range(3)])
+
+    return grad_B_vector_cylindrical
+
+def Bfield_gradient_cartesian(self):
+    '''
+    Function to calculate the gradient of the magnetic field vector B=(B_x,B_y,B_z)
+    at every point along the axis (hence with nphi points) where x, y and z
+    are the standard cartesian coordinates.
+    '''
+
+    B0, B1, B2 = self.Bfield_cylindrical()
+    nablaB = self.Bfield_gradient_cylindrical()
+    cosphi = np.cos(self.phi)
+    sinphi = np.sin(self.phi)
+    R0 = self.R0
+
+    grad_B_vector_cartesian = np.array([
+[cosphi**2*nablaB[0, 0] - cosphi*sinphi*(nablaB[0, 1] + nablaB[1, 0]) + 
+   sinphi**2*nablaB[1, 1], cosphi**2*nablaB[0, 1] - sinphi**2*nablaB[1, 0] + 
+   cosphi*sinphi*(nablaB[0, 0] - nablaB[1, 1]), cosphi*nablaB[0, 2] - 
+   sinphi*nablaB[1, 2]], [-(sinphi**2*nablaB[0, 1]) + cosphi**2*nablaB[1, 0] + 
+   cosphi*sinphi*(nablaB[0, 0] - nablaB[1, 1]), sinphi**2*nablaB[0, 0] + 
+   cosphi*sinphi*(nablaB[0, 1] + nablaB[1, 0]) + cosphi**2*nablaB[1, 1], 
+  sinphi*nablaB[0, 2] + cosphi*nablaB[1, 2]], 
+ [cosphi*nablaB[2, 0] - sinphi*nablaB[2, 1], sinphi*nablaB[2, 0] + cosphi*nablaB[2, 1], 
+  nablaB[2, 2]]
+    ])
+
+    return grad_B_vector_cartesian
+
+def Bfield_gradient_gradient_cylindrical(self):
```

Comment:
Consider renaming this function for consistency with the names `grad_B_tensor` and `grad_grad_B_tensor` used elsewhere in pyqsc.